### ISSUE TITLE
Could you please reveal more details on the training procedure?

### ISSUE THREAD

**Issue by SnowyJune973** [icoz69/DeepEMD#1]
今天我刚刚读过您（们）的文章，确实是很有启发性的思路（尤其是structured FC相关的讨论）！然而，文章并没有详细解释训练的细节，同时我注意到文章中使用的多个对比方法的结果都相当高（这些应该是您用ResNet-12作backbone自行复现的结果），故而我对您训练/推理时的具体流程产生了兴趣，能请您分享一些相关细节吗？谢谢！
关于“结果相当高的对比方法”：举例来说，表4a中使用ResNet-12作backbone的ProtoNe在 miniImagenet 5way 5shot上有78.01%，而根据我之前的实验结果，在不使用数据增强、并使用了预训练backbone的情况下训练20万个episode后，这个数值在71%左右；CTM中也提及了，用ResNet-18作backbone的ProtoNet在5way 5shot上的准确率在74%。

I've just read your paper and I have to say that it's really thoughtful and inspiring ( especially the discussions on the structured fully connected layer). However, the details on the training procedure aren't discussed in the paper; and I notice that the results of many previous state-of-the-arts (which are obtained by your reproduction, I suppose)  are relatively high in the paper than what I used to know, which makes me wonder how you managed to get them. Could you please reveal more details on your training/inference procedure? Thanks a lot.

About "relatively high in the paper": For example, in Table 4(a) ProtoNet(with ResNet-12 as its backbone) gets an accuracy of 78.01% on miniImagenet 5-way 5-shot tasks. However, according to my previous experiments, the performance of ProtoNet+ResNet12 is around 71% after 200,000 episodes of training, when a pretrained backbone is adopted and data augmentation is not used. It's also mentioned in CTM that ProtoNet+ResNet18 shows a performance of ~74% on miniImagenet 5-way 5-shot tasks.

**Comment by SnowyJune973** [comment]
emmm，CTM是您文章的引文28提出的

CTM is proposed in the 28th reference in your paper

**Comment by icoz69** [comment]
hello 
thank you very much for your interest. 
data augmentation is necessary,  below is the image transformations we use for data augmentation.

   
 if setname == 'val' or setname == 'test':
            # raw
            image_size = 84  # 80
            self.transform = transforms.Compose([
                transforms.Resize([92, 92]),
                transforms.CenterCrop(image_size),
                transforms.ToTensor(),
                transforms.Normalize(np.array([x / 255.0 for x in [125.3, 123.0, 113.9]]),
                                     np.array([x / 255.0 for x in [63.0, 62.1, 66.7]]))])
        elif setname == 'train':
            image_size = 84  # 80
            self.transform = transforms.Compose([
                transforms.RandomResizedCrop(image_size),
                transforms.RandomHorizontalFlip(),
                transforms.ToTensor(),
                transforms.Normalize(np.array([x / 255.0 for x in [125.3, 123.0, 113.9]]),
                                     np.array([x / 255.0 for x in [63.0, 62.1, 66.7]]))])

Both pre-training and episode training use learning decay, so you may try multiple parameters to get the best performance. Moreover, at the pre-training stage, the FC layer has a hidden layer. Setting a temperature for the category score is also important.
Our project is based on the open-source project of  [FEAT](https://github.com/Sha-Lab/FEAT). You can find more details there.
We are going to release our project code very soon.  
By the way, for questions regarding the paper, please drop an email to the **first author**. Github's issue is mainly for reporting coding problems. Thank you for your understanding.

**Comment by Franklin-Yao** [comment]
@SnowyJune973 check my code: https://github.com/Franklin-Yao/StrongProtoNet. You will get over 79%

**Comment by Harry-Up** [comment]
@icoz69 Hi, thanks for your work! I see the image size 80*80 is commented. I wonder if there are some differences in results between 84*84 and 80*80.

**Comment by icoz69** [comment]
> @icoz69 Hi, thanks for your work! I see the image size 80_80 is commented. I wonder if there are some differences in results between 84_84 and 80*80.

hello. i tried 80 80 in the early stage of this project and replace it to 84by84 to align it with many previous works. as far as I have experimented, the results are pretty close. you may try to modify it and do a comparison.

**Comment by ljjcoder** [comment]
@icoz69 Hi, thanks for your work! I noticed that the data set you used was the original image of miniimagenet, and then scaled to 84x84 by transforms.RandomResizedCrop(image_size). But miniimagenet itself should be 84x84 size. Have you ever compared, if the strict 84x84 miniimagenet data set is used, is it different from the data set you have transformed from the original size to 84x84?

**Comment by icoz69** [comment]
hello. i did not compare. I guess it may cause some performance drop, particularly for deepemd-grid and deepemd-sampling, as patches must  be  cropped from input images. so, i think it is better to crop before resize to 8484.

**Comment by ljjcoder** [comment]
@icoz69 Thank you for your prompt reply! To be fair, I think that all data augmentation should be performed on images that have been scaled to 84x84 like other few-shot learning methods. If data enhancement is performed on the original data, it should generate data that standard miniimagenet (84x84) cannot generate even through data enhancement.

**Comment by icoz69** [comment]
hello, we observe many SOTAmethods did so as well, but, yeah, I agree that it would be better if all the methods can be compared under an exactly same experimental configuration, such as same lr decay policy, same data Augmentation, same weight stabilization, etc..
in our paper, we re-implement many previous works for comparison in the experiment, such as  protonet, matchingnet , baseline++,baseline. 
The data augmentations we used are all common operations, which become standard. As far as I know, scale-then-crop is not very common in the classification papers.
Similarly,  early works do not apply pre-training and data augmentation, but they are standard in recent works.
